### PR TITLE
Added ability to highlight full text search results

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/android-lantern#310

- [x] Do the tests pass? Consistently?
- [x] Did this change improve test coverage?
- [ ] Has it been reviewed/approved by relevant teammates?
- [ ] Can it be QA’d in staging or something like staging?
- [ ] Is the code in question being linted? If not, consider adding a linter step to CI. If yes, make sure the linter is happy.
- [ ] Do we know how we’re going to deploy this?
- [ ] Are we clear on what the support and maintenance impact is?
- [ ] Did you create new documentation? Is it accessible and in the right place?
- [ ] Has existing documentation been updated?
- [ ] Have you logged tickets for related technical debt with the label “techdebt”?

### CODE REVIEW

File: db/src/androidTest/java/io/lantern/db/DBTest.kt [hunk 1]
```diff
@@ -197,7 +197,7 @@ class DBTest {
                     PathAndValue("/messages/c", Message("Message C blah blah")),
                     PathAndValue("/messages/d", Message("Message D blah blah blah"))
                 ),
-                db.list<String>("/messages/%")
+                db.list<Message>("/messages/%")
```
Comment (oxtoacart) [inline db/src/androidTest/java/io/lantern/db/DBTest.kt:200]: Drive-by, this test was using the wrong type, but it didn't break anything because we never tried accessing anything from that type.

File: db/src/main/java/io/lantern/db/DB.kt [hunk 2]
```diff
@@ -259,7 +259,7 @@ class DB private constructor(
         // Create a table for full text search
         db.execSQL(
             """CREATE VIRTUAL TABLE IF NOT EXISTS ${schema}_fts
-                        USING fts5(value, content=data, tokenize='porter unicode61')"""
+                        USING fts5(value, tokenize='porter trigram')"""
```
Comment (oxtoacart) [inline db/src/main/java/io/lantern/db/DB.kt:267]: Instead of using an external content table, we're using a regular fts5 table now.

To get prefix matching and highlighting, we need to use the experimental `trigram` tokenizer.

File: db/src/main/java/io/lantern/db/DB.kt [hunk 3]
```diff
@@ -993,32 +1040,14 @@ class Transaction internal constructor(
...
-    fun <T : Any> delete(path: String, extractFullText: ((T) -> String)? = null) {
+    fun delete(path: String) {
         val serializedPath = serde.serialize(path)
-        extractFullText?.let {
-            db.rawQuery(
-                "SELECT rowid, value FROM ${schema}_data WHERE path = ?", arrayOf(serializedPath)
-            ).use { cursor ->
-                if (cursor != null && cursor.moveToNext()) {
-                    db.execSQL(
-                        "INSERT INTO ${schema}_fts(${schema}_fts, rowid, value) VALUES('delete', ?, ?)",
```
Comment (oxtoacart) [inline db/src/main/java/io/lantern/db/DB.kt:1004]: Since we're using a normal fts5 table now, we don't need this special delete syntax anymore.

File: db/src/main/java/io/lantern/db/DB.kt [hunk 4]
```diff
@@ -993,32 +1040,14 @@ class Transaction internal constructor(
     /**
      * Deletes the value at the given path
      */
-    fun <T : Any> delete(path: String, extractFullText: ((T) -> String)? = null) {
```
Comment (oxtoacart) [inline db/src/main/java/io/lantern/db/DB.kt:996]: Another nice benefit of this change is that we don't need `extractFullText` in order to be able to delete from the index.

File: db/src/main/java/io/lantern/db/DB.kt [hunk 5]
```diff
@@ -256,10 +257,14 @@ class DB private constructor(
             """CREATE INDEX IF NOT EXISTS ${schema}_data_value_index ON ${schema}_data(value)
                         WHERE SUBSTR(CAST(value AS TEXT), 1, 1) = 'T'"""
         )
-        // Create a table for full text search
+        if (dropOldFullTextIndex) {
+            // Drop old version of full text search table
+            db.execSQL("DROP TABLE IF EXISTS ${schema}_fts")
```
Comment (oxtoacart) [inline db/src/main/java/io/lantern/db/DB.kt:262]: Since we never used to use full text indexing, it's okay to just drop the old table.

File: db/src/main/java/io/lantern/db/DB.kt [hunk 6]
```diff
@@ -745,12 +791,21 @@ open class Queryable internal constructor(
...
+        fullTextHighlightStartDelim: String = "*",
+        fullTextHighlightEndDelim: String = "*",
         onRow: (cursor: Cursor) -> Unit
     ) {
         val cursor = if (fullTextSearch != null) {
             db.rawQuery(
-                "SELECT d.path, d.value FROM ${schema}_fts f INNER JOIN ${schema}_data d ON f.rowid = d.rowid WHERE d.path LIKE ? AND f.value MATCH ? ORDER BY f.rank LIMIT ? OFFSET ?",
-                arrayOf(serde.serialize(pathQuery), fullTextSearch, count, start)
+                "SELECT d.path, d.value, highlight(${schema}_fts2, 0, ?, ?) FROM ${schema}_fts2 f INNER JOIN ${schema}_data d ON f.rowid = d.rowid WHERE d.path LIKE ? AND f.value MATCH ? ORDER BY f.rank LIMIT ? OFFSET ?",
+                arrayOf(
```
Comment (cosmicespresso) [inline db/src/main/java/io/lantern/db/DB.kt:801]: @oxtoacart Do you mind explaining the `ORDER BY f.rank`? (I was trying to find with what order logic the results come back to the client)

Comment (oxtoacart) [reply db/src/main/java/io/lantern/db/DB.kt:801]: Yes, f.rank is a virtual column that indicates how well the row matches the query. So what this means is that results are returned in the order of best to least best match.
